Add pipeline component examples catalogs for KFP and AA and MLX connector

ptitzler opened the pull request:
This PR 
- adds an example catalog connector for Kubeflow Pipelines 
- adds an example catalog connector for Apache Airflow
- adds a catalog connector for the Machine Learning Exchange
- ~requires https://github.com/elyra-ai/elyra/pull/2263~ **RESOLVED**
- ~requires https://github.com/elyra-ai/elyra/pull/2267 (`+` > `New undefined` shown until merged)~ **RESOLVED**
- ~requires https://github.com/elyra-ai/elyra/pull/2282~ **RESOLVED**
- adds a catalog connector directory
- adds a build-your-own-connector stub
- updates the repository README with links to the connector content

**Add examples catalog user experience (Airflow shown):**
- Open "Manage Components"
- Add catalog (`+`)
- `New Apache Airflow example components catalog`
- Enter a catalog name (any string) - all other inputs are pre-populated.
   ![image](https://user-images.githubusercontent.com/13068832/140975460-ca84722a-86ab-4c5b-bcbf-dcd30057992c.png)

  Note that the **Learn more...** link opens the connector's GitHub repository README, which contains information about the connector. (For built-in catalog types the link opens the pipeline components topic in the user guide, which introduces catalogs.)

- Expand palette in VPE
   ![image](https://user-images.githubusercontent.com/13068832/140976162-2d0d471b-3782-4058-a0be-5ca0f103c1ce.png)


Note that the following information is persisted for each node in the `.pipeline` file (irrelevant information was removed from the KFP snippet below):
```
"type": "execution_node",
"op": "elyra-kfp-examples-catalog:737915b826e9",
"component_description": "Filter input text according to the given regex pattern using shell and grep.",
"component_source": "{'catalog_type': 'elyra-kfp-examples-catalog', 'component_ref': {'component-id': 'filter_text_using_shell_and_grep.yaml'}}",
```

Airflow:
```
"type": "execution_node",
"op": "elyra-airflow-examples-catalog:3a55d015ea96",
"component_source": "{'catalog_type': 'elyra-airflow-examples-catalog', 'component_ref': {'component-id': 'bash_operator.py'}}",
```
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

kevin-bates commented on component-catalog-connectors/example-components-connector/README.md:
```diff
@@ -0,0 +1,52 @@
+1. [Open the '`Manage Components`' panel](
+https://elyra.readthedocs.io/en/stable/user_guide/pipeline-components.html#managing-custom-components-using-the-jupyterlab-ui).
+1. Add a new component examples catalog ('`+`' > '`New Elyra component examples catalog`').
+1. Specify a catalog name, e.g. '`Elyra example components for Kubeflow`'.
```
What is the behavior when a second instance of the same runtime is created?  Do users just see another set of operators?

Should there be a way to filter out components?  Could this be done by exposing the `file_filter` property in the schema?  (the problem there is that the default value would probably need to be `*` since it needs to span the runtime types, although it looks like kfp only has `yaml` and airflow only has `python`.)

kiersten-stokes replied:
I think since the relative file paths would be the same for each instance, there won't be any duplication of components/nodes/operators because, if anything, the ones read/processed second will 'overwrite' (with the same information) the ones read/processed first

kevin-bates replied:
Oh right.  I was thinking its the connector instance name that is used in the hash, but it's really the instance's underlying schema name.  (Is that correct?  Is the hashed op value <schema_name>:<hash of provided key values>?)

Regarding the filtering aspect, should users be able to filter on the component definition name?  This would allow them to distribute components across different categories.

ptitzler replied:
> (Is that correct? 

Yes!


> Regarding the filtering aspect

Several reasons why I'm leaning towards not supporting this (_for example connectors_ - but I am agreeing that this makes sense for generic packaged connectors):
 - We have a very small set of components to begin with.
 - Users won't know ahead of time what components are included and thus what tp filter by.

Also, we  really don't want to encourage users to use these examples for anything other than initial exploration. This was one concern I had with our initial approach: users must not assume that the same example components are available in _every_ Elyra release. For example, if we were to remove the spark operators in the future, all "production" pipelines that relied on those operators would fail. Users should therefore always only use components from catalogs that they have control over.

kevin-bates commented on component-catalog-connectors/example-components-connector/elyra_examples_connector/examples_connector.py:
```diff
@@ -0,0 +1,117 @@
+        """
+        component_list = []
+
+        runtime_id = catalog_metadata.get('runtime')
```
Could we update `runtime_id` to `runtime_type` and not introduce another incantation of the same thing?
```suggestion
        runtime_type = catalog_metadata.get('runtime')
```
We'll also need to update the field to `runtime_type` with the RuntimeProcessorType name values throughout.

ptitzler replied:
done (variable names and message text)

lresende commented:
Please make sure to update LICENSE files accordingly to mention new files as in [PR 2272](https://github.com/elyra-ai/elyra/pull/2272/files#diff-c693279643b8cd5d248172d9c22cb7cf4ed163a3c98c8a3f69c2717edd3eacb7R245)

kiersten-stokes reviewed (approved):
This is working great! I installed from source for my tests

ptitzler commented on component-catalog-connectors/example-components-connector/elyra_examples_connector/examples_connector.py:
```diff
@@ -0,0 +1,117 @@
+            pattern = ExamplesCatalogConnector.config[runtime_id].get('file_filter', '*')
+            self.log.debug(f'Component file pattern: {pattern}')
+            for file in root_dir.glob(f'**/{pattern}'):
+                component_list.append({'component-id': str(file)[len(str(root_dir)) + 1:]})
```
Yep - that was the intention. Since this connector serves multiple purposes (delivers example components + blueprint for other connectors) I wanted to keep the key name generic.

ptitzler commented:
A couple pending changes:
- split connector into two connectors (one for kfp, one for airflow):
  -  avoids issues like the ones we are facing with other BYOs (stuff that doesn't belong together is stored together) -> this will make it easier for somebody to create an "examples" connector for another runtime type
  - customized schema, which pre-populates as much as possible, e.g. for KFP
   - ![image](https://user-images.githubusercontent.com/13068832/140806115-83903ef7-e54c-4948-9a87-06d526f50066.png)
  -  ![image](https://user-images.githubusercontent.com/13068832/140806637-938d6227-e8d0-42e7-be0c-bd81db6a6957.png)


Note:
- the elyra `setup.py` will list both connectors as extras
- the **`Dockerfile` templates** we include in the elyra repo will install [the relevant] connector package(s) AND invoke `elyra-metadata install ...`. This way the example components are available in the VPE palette by default. Since these are just templates, customers can choose to remove the dependencies/setup steps in their customized version if they want to exclude the example components (or use their own "forked" examples instead.

kevin-bates commented:
The updates sound great.  I still think it would be good to expose a filter/regex schema property that can be applied so users can build connector instances that are subsets of the complete set.

ptitzler commented:
> Please make sure to update LICENSE files accordingly to mention new files as in [PR 2272](https://github.com/elyra-ai/elyra/pull/2272/files#diff-c693279643b8cd5d248172d9c22cb7cf4ed163a3c98c8a3f69c2717edd3eacb7R245)

Thank you, I have totally missed that. I added a license file to the Kubeflow Pipelines connector's `resources` directory. The Airflow components already have their own license embedded, so we should be good there!

ptitzler commented on component-catalog-connectors/example-components-connector/README.md:
````diff
@@ -0,0 +1,52 @@
+
+   ```
+   $ git clone https://github.com/elyra-ai/examples.git
+   $ cd examples/component-catalog-connectors/elyra-examples-connector/
````
FIxed!

kevin-bates commented:
Sorry - I just noticed my comments were applied to a specific commit, which requires you to have to open the commit link to get their context.  What a hassle - I apologize.

kevin-bates commented on component-catalog-connectors/airflow-example-components-connector/README.md:
```diff
@@ -0,0 +1,60 @@
+
+**Prerequisites**
+
+- [Install Elyra](https://elyra.readthedocs.io/en/stable/getting_started/installation.html) (version 3.3 and above).
```
Won't this include a dependency on the examples and prevent the need for the `pip install` below?

Hmm - or are we not doing that since this is now split by runtime-type?

ptitzler replied:
My main motivation was to explicitly highlight that version 3.3 is required and ...

You are right that

```
pip install elyra[all]
```
(pending PR) will install the examples (for KFP and AA). What is not clear is if users actually install that way or omit `[all]`. In the latter case, an explicit install would still be required. In a way I am trying to cover all bases without if then else (:

Once we have runtime specific targets (e.g. `elyra[kfp-only]`) only the runtime specific examples will be installed.

kevin-bates replied:
> Once we have runtime specific targets (e.g. elyra[kfp-only]) only the runtime specific examples will be installed.

I personally think we're going to need to move to a mono-repo that supports elyra, elyra-kfp, and elyra-airflow packages since I don't think the `extra_requires` approach will handle entry_points in the manner we need.

kevin-bates commented on component-catalog-connectors/airflow-example-components-connector/README.md:
```diff
@@ -0,0 +1,60 @@
+1. [Open the '`Manage Components`' panel](
+https://elyra.readthedocs.io/en/stable/user_guide/pipeline-components.html#managing-custom-components-using-the-jupyterlab-ui).
+1. Add a new component examples catalog ('`+`' > '`New Apache Airflow example components catalog`').
+1. Specify a catalog name, e.g. '`Elyra example components for Kubeflow`'.
```
Did you mean to reference Kubeflow here?
```suggestion
1. Specify a catalog name, e.g. '`Elyra example components for Airflow`'.
```

ptitzler replied:
Copy and paste error. Fixed!

kevin-bates commented on component-catalog-connectors/airflow-example-components-connector/setup.py:
```diff
@@ -38,7 +38,7 @@
     packages=find_packages(),
     install_requires=[
-        'elyra==3.3.0.dev0'
+        'elyra>3.2'
```
If elyra has a dependency on this, will this trigger a circular dependency?  I suppose for elyra, the plan is to use an `extras_require` to pull in the examples: `pip install elyra[examples]` or `pip install elyra[examples-kfp]` - so we're probably okay.

ptitzler replied:
I removed the dependency

kevin-bates reviewed (commented):
Aside from the circular ref comment below, the connector schemas should include the following updates (can't apply in-line commenting so adding here):
- The `$id` values still reflect the combined form of the repo
- The `display_name` property should be removed as it's redundant with `title`.  (This comment is NOT apply to the `display_name` property definition down in the `properties` stanza - which applies to _instances_.)  All schemas are being updated in this manner in https://github.com/elyra-ai/elyra/pull/2267
- The `schemaspace` and `schemaspace_id` values should be updated to `component-catalogs` and `8dc89ca3-4b90-41fd-adb9-9510ad346620`, respectively once https://github.com/elyra-ai/elyra/pull/2282 is merged.

ptitzler commented:
> [...] the connector schemas should include the following updates (can't apply in-line commenting so adding here):

> * The `$id` values still reflect the combined form of the repo

Fixed in both schemas!

> * The `display_name` property should be removed as it's redundant with `title`.  (This comment is NOT apply to the `display_name` property definition down in the `properties` stanza - which applies to _instances_.)  All schemas are being updated in this manner in [elyra-ai/elyra#2267](https://github.com/elyra-ai/elyra/pull/2267)

Removed the top level `display_name` property from both schemas!

> * The `schemaspace` and `schemaspace_id` values should be updated to `component-catalogs` and `8dc89ca3-4b90-41fd-adb9-9510ad346620`, respectively once [elyra-ai/elyra#2282](https://github.com/elyra-ai/elyra/pull/2282) is merged.

Adding this to the TODO list

kiersten-stokes reviewed (approved):
LGTM! I incorporated the changes that this PR is dependent on and tested all three catalog types -- all are working great

ptitzler commented:
> * `chemaspace` and `schemaspace_id` values should be updated to `component-catalogs` and `8dc89ca3-4b90-41fd-adb9-9510ad346620`, respectively once [elyra-ai/elyra#2282](https://github.com/elyra-ai/elyra/pull/2282) is merged.

Done!

ptitzler commented:
@akchinSTC @lresende @kevin-bates @kiersten-stokes you should have received  pending collaborator invitations for three packages from PyPI

ptitzler commented:
For test purposes version 0.0.1 of each package was published to unblock https://github.com/elyra-ai/elyra/pull/2286. Once development has completed and no issues were found version 0.0.1 will be replaced by 0.1.0.
